pkg/common/plugin/httpchallenge/httpchallenge.go

```diff
@@ -0,0 +1,126 @@
+package httpchallenge
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+	"github.com/spiffe/spire/pkg/common/idutil"
+)
+
+const (
+	nonceLen = 32
+
+	// PluginName for http based attestor
+	PluginName = "http_challenge"
+)
+
+var (
+	DefaultRandReader io.Reader
+)
+
+type AttestationData struct {
+	HostName  string `json:"hostname"`
+	AgentName string `json:"agentname"`
+	Port      int    `json:"port"`
+}
+
+type Challenge struct {
+	Nonce string `json:"nonce"`
+}
+
+type Response struct {
+}
+
+func GenerateChallenge() (*Challenge, error) {
+	nonce, err := generateNonce()
+	if err != nil {
+		return nil, err
+	}
+	return &Challenge{Nonce: nonce}, nil
+}
+
+func CalculateResponse(_ *Challenge) (*Response, error) {
+	return &Response{}, nil
+}
+
+func VerifyChallenge(attestationData *AttestationData, challenge *Challenge) error {
+	if attestationData.HostName == "" {
+		return fmt.Errorf("hostname must be set")
+	}
+	if attestationData.AgentName == "" {
+		return fmt.Errorf("agentname must be set")
+	}
+	if attestationData.Port <= 0 {
+		return fmt.Errorf("port is invalid")
+	}
+	if strings.Contains(attestationData.HostName, "/") {
+		return fmt.Errorf("hostname can not contain a slash")
+	}
+	if strings.Contains(attestationData.HostName, ":") {
+		return fmt.Errorf("hostname can not contain a colon")
+	}
+	if strings.Contains(attestationData.AgentName, ".") {
+		return fmt.Errorf("agentname can not contain a dot")
+	}
+	turl := url.URL{
+		Scheme: "http",
+		Host:   net.JoinHostPort(attestationData.HostName, strconv.Itoa(attestationData.Port)),
+		Path:   fmt.Sprintf("/.well-known/spiffe/nodeattestor/http_challenge/%s/challenge", attestationData.AgentName),
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```

Comment:
This timeout is probably better applied by the caller. And instead of using context.Background(), it should use the context from the RPC so that it respects stream cancellation.

Reply:
Do you mean, the value of the timeout should be passed to the function, or the whole context.WithTimeout should be lifted out, and its resulting context passed to this function?

Reply:
I mean the latter. The plugin should pass its stream context to this function. The benefit there is that if the RPC is cancelled ahead of the timeout then the plugin can fail verification immediately and return. The plugin can set the timeout on the context before passing it into this function.

Reply:
k. I think I implemented this. but not sure if its doing the right thing with context.Background() still...